index.js

```diff
@@ -4,61 +4,78 @@ const childProcess = require('child_process');
-const xprop = 'xprop -id $(xprop -root 32x \'\\t$0\' _NET_ACTIVE_WINDOW | cut -f 2)';
+const xpropActive = 'xprop -root 32x \'\\t$0\' _NET_ACTIVE_WINDOW | cut -f 2';
+const xpropDetails = 'xprop -id';
 
-function parse(stdout) {
-	if (process.platform === 'darwin') {
-		const parts = stdout.trimRight().split('\n');
+const parseMac = stdout => {
+	const parts = stdout.trimRight().split('\n');
 
-		return {
-			title: parts[0],
-			id: Number(parts[1]),
-			app: parts[2],
-			pid: Number(parts[3])
-		};
-	} else if (process.platform === 'linux') {
-		const result = {};
+	return {
+		title: parts[0],
+		id: Number(parts[1]),
+		app: parts[2],
+		pid: Number(parts[3])
+	};
+};
 
-		for (const row of stdout.trim().split('\n')) {
-			if (row.includes('=')) {
-				const parts = row.split('=');
-				result[parts[0].trim()] = parts[1].trim();
-			} else if (row.includes(':')) {
-				const parts = row.split(':');
-				result[parts[0].trim()] = parts[1].trim();
-			}
-		}
-		const windowIdProperty = 'WM_CLIENT_LEADER(WINDOW)';
-		const resultKeys = Object.keys(result);
-		const windowId = (resultKeys.indexOf(windowIdProperty) > 0 &&
-			parseInt(result[windowIdProperty].split('#').pop(), 16)) || null;
+const parseLinux = linuxData => {
+	const stdout = linuxData.stdout;
+	const activeWindowId = linuxData.activeWindowId;
 
-		return {
-			title: JSON.parse(result['_NET_WM_NAME(UTF8_STRING)']) || null,
-			id: windowId,
-			app: JSON.parse(result['WM_CLASS(STRING)'].split(',').pop()),
-			pid: parseInt(result['_NET_WM_PID(CARDINAL)'], 10)
-		};
+	const result = {};
+	for (const row of stdout.trim().split('\n')) {
+		if (row.includes('=')) {
+			const parts = row.split('=');
+			result[parts[0].trim()] = parts[1].trim();
+		} else if (row.includes(':')) {
+			const parts = row.split(':');
+			result[parts[0].trim()] = parts[1].trim();
+		}
 	}
 
-	throw new Error('macOS and Linux only');
-}
+	const windowIdProperty = 'WM_CLIENT_LEADER(WINDOW)';
+	const resultKeys = Object.keys(result);
+	const windowId = (resultKeys.indexOf(windowIdProperty) > 0 &&
+		parseInt(result[windowIdProperty].split('#').pop(), 16)) || activeWindowId;
+
+	return {
+		title: JSON.parse(result['_NET_WM_NAME(UTF8_STRING)']) || null,
+		id: windowId,
+		app: JSON.parse(result['WM_CLASS(STRING)'].split(',').pop()),
+		pid: parseInt(result['_NET_WM_PID(CARDINAL)'], 10)
+	};
+};
 
 module.exports = () => {
 	if (process.platform === 'darwin') {
-		return pify(childProcess.execFile)(bin).then(parse);
+		return pify(childProcess.execFile)(bin).then(parseMac);
 	} else if (process.platform === 'linux') {
-		return pify(childProcess.exec)(xprop).then(parse);
+		return pify(childProcess.exec)(xpropActive).then(
+			activeWindowIdStdout => {
+				const activeWindowId = activeWindowIdStdout.trim();
+				return pify(childProcess.exec)(`${xpropDetails} ${activeWindowId}`).then(stdout => {
+					return {
+						activeWindowId,
+						stdout
+					};
+				});
+			}
+		).then(parseLinux);
 	}
```

Comment:
I'm not completely sure about the code "repetition" in here and in the `sync` version below.

I first had coded it as a single separated function, but it depended on using Promises, so I had to separate it in these two versions. As now, for Linux, Node would be calling 2 independent processes (first to get the ID and then to get the details). So, the structure of the code changes a bit in the async and sync version.

I also thought about finding and adding as a dependency another package to return the resolved value of the promises in the `sync` version (instead of a "Promise" / "thenable"). And then just take care of the Promise based async code. But I thought you probably wouldn't like to add another dependency just for that (I don't even know if there's a package to achieve that).

What do you think @sindresorhus ?